src/verify_expr.ml

```diff
@@ -2049,32 +2105,63 @@ module VerifyExpr(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
+              List.map (fun (sign,info) -> (List.map erase_type sign, info)) cctors in
+            List.filter (fun (sign, _) -> is_assignable_to_sign (Some pure) argtps_erased sign) cctors 
+      in
       begin match consmap' with
         [] -> static_error l "No matching constructor" None
       | [(sign, CtorInfo (lm, xmap, pre, pre_tenv, post, epost, terminates, ss, v))] ->
-        let obj = get_unique_var_symb (match xo with None -> "object" | Some x -> x) (ObjType cn) in
+        (* Typecheck the type arguments *)
+        let targtps = match targs with 
+          (* Type check passed type arguments *)
+          Some(targs) -> List.map (check_pure_type (pn,ilist) tparams Real) targs 
+          (* Raw type, make all targs Object *)
+          | None -> List.map (fun _ -> javaLangObject) tparams in
+        let obj = get_unique_var_symb (match xo with None -> "object" | Some x -> x) (ObjType (cn, targtps)) in
         assume_neq obj (ctxt#mk_intlit 0) $. fun () ->
         assume_eq (ctxt#mk_app get_class_symbol [obj]) (List.assoc cn classterms) $. fun () ->
         let is_upcall =
           match ss, leminfo with
             Some (Some (_, rank)), RealMethodInfo (Some rank') -> rank < rank'
           | _ -> true
         in
+        let xmap = List.map (fun (name,tp) -> (name, erase_type tp)) xmap in
         check_correct h None None [] args (lm, [], None, xmap, ["this", obj], pre, post, Some(epost), terminates, Static) is_upcall (Some cn) (fun h env _ -> cont h env obj)
       | _ -> static_error l "Multiple matching overloads" None
       end
-    | WMethodCall (l, tn, m, pts, args, fb) when m <> "getClass" ->
+    | WMethodCall (l, tn, m, pts, args, fb, tpenv) when m <> "getClass" ->
+      let get_assignable_methods meths = List.filter (fun ((name,params), info) -> name = m && is_assignable_to_sign (if pure then Some(true) else None) pts params) meths in
       let (lm, gh, rt, xmap, pre, post, epost, terminates, is_upcall, target_class, fb', v) =
         match try_assoc tn classmap with
-          Some {cfinal; cmeths} ->
-          let MethodInfo (lm, gh, rt, xmap, pre, pre_tenv, post, epost, pre_dyn, post_dyn, epost_dyn, terminates, ss, fb, v, is_override, abstract) = List.assoc (m, pts) cmeths in
+        Some {cfinal; cmeths} ->
+          let MethodInfo (lm, gh, rt, xmap, pre, pre_tenv, post, epost, pre_dyn, post_dyn, epost_dyn, terminates, ss, fb, v, is_override, abstract) = 
+            try List.assoc (m, pts) cmeths with 
+            Not_found -> (* Method contains type parameters, so apply erasure *)
+              let cmeths = List.map (fun ((name,params), info) -> ((name, List.map erase_type params), info)) cmeths in
+              match get_assignable_methods cmeths with
+                [] -> static_error l "No matching method for call." None
+                | [m] -> snd m
+                | _ -> static_error l "Multiple matching matchods for call." None
+          in

```

Comment:
Should not be necessary; make sure in `verifast1.ml` that the `WMethodCall` contains the exact `pts` from the `cmeths` map. (The only purpose of `pts` is to be able to look up the method in `cmeths`!)